Processing port geometry tools

alexbruy opened the pull request:
## Description
Port some Python algorithms to C++ as a part of [QEP #271](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/271).

Voronoi polygons and Delaunay triangulation algorithms ported to C++ and use GEOS to compute operations. Old Python code removed.

Concave hull algorithm ported to C++, old Python code removed. K-nearest concave hull marked as deprecated, but still available to keep existing scripts/models working.

alexbruy committed:
[processing] port voronoi polygons algorithm to C++

alexbruy committed:
[processing] port delaunay triangulation algorithm to C++

alexbruy committed:
[processing] port concave hull algorithm to C++

agiudiceandrea commented:
Hi @alexbruy, does this also addresses the issue reported at About the https://github.com/qgis/QGIS/issues/52172?

alexbruy committed:
try to fix tests

alexbruy committed:
another attempt to fix test

nyalldawson commented on src/analysis/processing/qgsalgorithmvoronoipolygons.cpp:
```diff
@@ -0,0 +1,194 @@
+  long i = 0;
+  double step = source->featureCount() > 0 ? 50.0 / source->featureCount() : 1;
+
+  const QgsSpatialIndex index( it, [&]( const QgsFeature & f )->bool
```
This only needs to happen if copyAttributes is true. Otherwise we can skip the cost of creating the index.

And if copyAttributes is true then we could potentially use https://github.com/libgeos/geos/pull/781 to avoid the index for newer geos versions

nyalldawson commented on src/analysis/processing/qgsalgorithmvoronoipolygons.cpp:
```diff
@@ -0,0 +1,194 @@
+        engine->prepareGeometry();
+        for ( const QgsFeatureId id : intersected )
+        {
+          if ( engine->intersects( index.geometry( id ).constGet() ) )
```
This should be implied when querying the index built using single points -- we could save the cost here and also avoid storing the geometries in the index.

Potentially a big speed boost by using QgsSpatialIndexKDBush here too.

alexbruy replied:
Not sure I understand. We can't check for exact intersection with the geometry, index supports only intersection with the rectangle, so we need test whether point actually intersets geometry.

nyalldawson commented on src/analysis/processing/qgsalgorithmconcavehull.cpp:
```diff
@@ -0,0 +1,250 @@
+void QgsConcaveHullAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Input layer" ), QList< int >() << QgsProcessing::TypeVectorPoint ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "ALPHA" ), QObject::tr( "Threshold" ), QgsProcessingParameterNumber::Double, 0.3, false, 0, 1 ) );
```
The extra "Threshold (0-1, where 1 is equivalent with Convex Hull)" help text could be added as the help for this parameter

nyalldawson commented:
Regarding concave hull -- can we ifdef in a geos version check and delegate the whole algorithm to geos on newer versions?

alexbruy commented:
> Regarding concave hull -- can we ifdef in a geos version check and delegate the whole algorithm to geos on newer versions?

I might be wrong, but from my tests geos implementation creates concave hull using different approach. At least for me, not matter which threshold I use it always produces the same polygon covering all points.

alexbruy committed:
more verbose label for parameter

alexbruy committed:
use geos to calculate concave hull on systems with geos >=3.11.0

alexbruy committed:
do not create spatial index if preserving attributes is not needed

alexbruy committed:
make cppcheck happy

alexbruy commented:
I may be wrong, but seems old Python implementation produces incorrect results by not including all input points into concave hull polygon.

uclaros commented on src/analysis/processing/qgsalgorithmvoronoipolygons.cpp:
```diff
@@ -0,0 +1,250 @@
+  delta = extent.height() * mBuffer / 100.0;
+  extent.setYMinimum( extent.yMinimum() - delta );
+  extent.setYMaximum( extent.yMaximum() + delta );
+  QgsGeometry clippingGeom = QgsGeometry::fromRect( extent );
+
+  QgsGeometry voronoiDiagram = allPoints.voronoiDiagram( clippingGeom, mTolerance );
```
```suggestion
  const QgsGeometry clippingGeom = QgsGeometry::fromRect( extent );

  const QgsGeometry voronoiDiagram = allPoints.voronoiDiagram( clippingGeom, mTolerance );
```

alexbruy replied:
Done.

uclaros reviewed (commented):
Nice! Just some minor comments

alexbruy committed:
address review

alexbruy commented on src/analysis/processing/qgsalgorithmdelaunaytriangulation.cpp:
```diff
@@ -0,0 +1,172 @@
+void QgsDelaunayTriangulationAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Input layer" ), QList< int >() << QgsProcessing::TypeVectorPoint ) );
+  addParameter( new QgsProcessingParameterNumber( QStringLiteral( "TOLERANCE" ), QObject::tr( "Tolerance" ), QgsProcessingParameterNumber::Double, 0, true, 0 ) );
```
Added

alexbruy commented on src/analysis/processing/qgsalgorithmdelaunaytriangulation.cpp:
```diff
@@ -0,0 +1,172 @@
+  QgsGeometry allPoints;
+
+  long i = 0;
+  double step = source->featureCount() > 0 ? 50.0 / source->featureCount() : 1;
```
Fixed, thanks!

alexbruy commented on src/analysis/processing/qgsalgorithmconcavehull.cpp:
```diff
@@ -0,0 +1,331 @@
+      allPoints.addPart( qgsgeometry_cast< QgsPoint * >( geom )->clone(), Qgis::GeometryType::Point );
+    }
+  }
+  QgsGeometry concaveHull = allPoints.concaveHull( mPercentage, mAllowHoles );
```
Done.

alexbruy commented on src/analysis/processing/qgsalgorithmdelaunaytriangulation.cpp:
```diff
@@ -0,0 +1,172 @@
+
+  if ( !triangulation.isEmpty() )
+  {
+    QVector< QgsGeometry > collection = triangulation.asGeometryCollection();
```
Done.

alexbruy commented on src/analysis/processing/qgsalgorithmdelaunaytriangulation.cpp:
```diff
@@ -0,0 +1,172 @@
+    return true;
+  }, QgsSpatialIndex::FlagStoreFeatureGeometries );
+
+  QgsGeometry triangulation = allPoints.delaunayTriangulation( tolerance );
```
Done.

alexbruy commented on src/analysis/processing/qgsalgorithmconcavehull.cpp:
```diff
@@ -0,0 +1,331 @@
+  long i = 0;
+  double step = mSource->featureCount() > 0 ? 50.0 / mSource->featureCount() : 1;
+
+  QgsFeatureIterator it = mSource->getFeatures( QgsFeatureRequest().setSubsetOfAttributes( QList< int >() ), QgsProcessingFeatureSource::FlagSkipGeometryValidityChecks );
```
Fixed.

alexbruy commented on src/analysis/processing/qgsalgorithmconcavehull.cpp:
```diff
@@ -0,0 +1,331 @@
+
+  if ( mSplitMultipart && concaveHull.isMultipart() )
+  {
+    QVector< QgsGeometry > collection = concaveHull.asGeometryCollection();
```
Done.

alexbruy commented on src/analysis/processing/qgsalgorithmvoronoipolygons.cpp:
```diff
@@ -0,0 +1,250 @@
+
+QString QgsVoronoiPolygonsAlgorithm::voronoiWithAttributes( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  QgsFields fields = mSource->fields();
```
Done.

alexbruy commented on src/analysis/processing/qgsalgorithmvoronoipolygons.cpp:
```diff
@@ -0,0 +1,250 @@
+  {
+    std::unique_ptr< QgsGeometryEngine > engine;
+    std::unique_ptr< QgsGeometryEngine > extentEngine( QgsGeometry::createGeometryEngine( clippingGeom.constGet() ) );
+    QVector< QgsGeometry > collection = voronoiDiagram.asGeometryCollection();
```
Done.

uclaros commented on src/analysis/processing/qgsalgorithmdelaunaytriangulation.cpp:
```diff
@@ -0,0 +1,174 @@
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Input layer" ), QList< int >() << QgsProcessing::TypeVectorPoint ) );
+  std::unique_ptr<QgsProcessingParameterNumber> toleranceParam = std::make_unique<QgsProcessingParameterNumber>( QStringLiteral( "TOLERANCE" ), QObject::tr( "Tolerance" ), QgsProcessingParameterNumber::Double, 0, true, 0 );
+  toleranceParam->setHelp( QObject::tr( "Specifies an optional snapping tolerance which can be used to improve the robustness of the triangulation" ) );
```
Wow! I didn't know you could do this! Why is it only used on 6 algorithms? This should be everywhere!

nyalldawson replied:
It's relatively new, introduced after the bulk of the algorithms were already made

nyalldawson replied:
But yes, should be everywhere 😁

alexbruy committed:
concave hull tests based on geos version

alexbruy committed:
use qobject_cast

alexbruy committed:
use run() to execute nested algorithms
move step calculation to prepareAlgorithm to avoid crash
disable tests for native implementation

alexbruy committed:
add test files

nirvn commented:
Nice one @alexbruy

DelazJ commented on src/analysis/processing/qgsalgorithmdelaunaytriangulation.cpp:
```diff
@@ -0,0 +1,174 @@
+  const double tolerance = parameterAsDouble( parameters, QStringLiteral( "TOLERANCE" ), context );
+  const bool addAttributes = parameterAsBool( parameters, QStringLiteral( "ADD_ATTRIBUTES" ), context );
+
+  QgsFields fields;
+  if ( addAttributes )
+  {
+    fields.append( QgsField( QStringLiteral( "POINTA" ), QVariant::LongLong ) );
+    fields.append( QgsField( QStringLiteral( "POINTB" ), QVariant::LongLong ) );
+    fields.append( QgsField( QStringLiteral( "POINTC" ), QVariant::LongLong ) );
+  }
+  else
+  {
+    fields.append( QgsField( QStringLiteral( "id" ), QVariant::LongLong ) );
+  }
```
@alexbruy I'm on my way to document this PR and wonder if it is deliberate that the output layer contains either an id field or the IDs from the input point. Why not always have the generated id field in the output layer (and add or not the original points)? I guess, something like:
```
  QgsFields fields;
  fields.append( QgsField( QStringLiteral( "id" ), QVariant::LongLong ) );
  if ( addAttributes )
  {
    fields.append( QgsField( QStringLiteral( "POINTA" ), QVariant::LongLong ) );
    fields.append( QgsField( QStringLiteral( "POINTB" ), QVariant::LongLong ) );
    fields.append( QgsField( QStringLiteral( "POINTC" ), QVariant::LongLong ) );
  }
```
Same applies to Voronoi afaict. Thanks.

alexbruy replied:
It is intentional, this is behaviour of the original Python implementation.
